### DISCUSSION TITLE
[FE210] 從新手到中手：前端工程加強班，Twitch API 更新

### DISCUSSION THREAD

**Discussion by a1740942002** [Lidemy/forum#44] [前端相關問題] [answered]
老師好，想請問關於 [FE210] 中第四單元：**4. 從假資料到真資料：Ajax 與 API 串接**，的部分。

由於 Twitch 的 API 有進行更新，所以 Get Stream API 的部分並沒有包含到實況主的個人大頭貼，而是拆出來分裝到 Get User API 之中。

因此不太確定我現在這樣的解法行不行，我是利用嵌套 callback 再 callback 的方式來達成的，這樣做好像就是所謂俗稱的回呼地獄吧？

例如：

```javascript
// 初始化
initStreams();

function initStreams(){
  let queryString = '?language=zh';

  //////////// 這邊開始回呼地獄 ////////////////

  getStreams(queryString, function(streams){
    streams.forEach(stream => { queryString += '&id=' + stream.user_id})

    getUsers(queryString, function(users){
      streams.forEach(stream => {
        users.forEach(user => {
          if(stream.user_id === user.id){
            stream.profile_image_url = user.profile_image_url
            stream.offline_image_url = user.offline_image_url
          }
        })
      })
      render(streams);
    });
  });
}

// API

function getStreams(queryString, callback){
  const api = 'https://api.twitch.tv/helix/streams';
  const clientID = 'kfwda6tzypba70e78x0fbntpnm59fc'; // Twitch 官網註冊
  const appToken = '35spin2autlycym825e63wr6990lzs'; // 註冊之後發 POST 取得 app access token
  const request = new XMLHttpRequest();
  request.open('GET', api + queryString, true);
  request.setRequestHeader('Authorization', 'Bearer ' + appToken);
  request.setRequestHeader('Client-Id', clientID);
  request.send();
  request.onload = function(e){
    const results = JSON.parse(request.response).data;
    callback(results);
  }
}

function getUsers(queryString, callback){
  const api = 'https://api.twitch.tv/helix/users';
  const clientID = 'kfwda6tzypba70e78x0fbntpnm59fc'; // Twitch 官網註冊
  const appToken = '35spin2autlycym825e63wr6990lzs'; // 註冊之後發 POST 取得 app access token
  const request = new XMLHttpRequest();
  request.open('GET', api + queryString, true);
  request.setRequestHeader('Authorization', 'Bearer ' + appToken);
  request.setRequestHeader('Client-Id', clientID);
  request.send();
  request.onload = function(e){
    const results = JSON.parse(request.response).data;
    callback(results);
  }
}
```

我記得這樣做不是很好，但如果不嵌套 callback function 的話，我不知道該如何取得實況主的大頭貼等資料。
( 上網查資料之後，發現好像可以使用 ES6 新出的 promise 進行解決，我每堂都有上，不過發現課程好像沒提到 promise？所以就先以 ES5 的方法進行解決了 )。

以下是我的作業與參考資料：

1. [原始碼](https://github.com/a1740942002/my_twitch_app/blob/promise/index.html)
2. [展示](https://a1740942002.github.io/my_twitch_app/)
3. [Twitch 新的 API](https://dev.twitch.tv/docs/api/reference)

> 註：作業的 CSS 部分是使用 TailwindCSS 框架進行撰寫，所以 HTML 中 class 會比較多。

**Comment by aszx87410** [comment, marked as answer]
> 因此不太確定我現在這樣的解法行不行，我是利用嵌套 callback 再 callback 的方式來達成的，這樣做好像就是所謂俗稱的回呼地獄吧？

這樣完全可以喔，你這才兩層而已，稱不上是回呼地獄，還差得遠呢XD  
至少要三層以上我覺得可讀性才會比較差一點，兩層真的還好，不過你底下 users.forEach 那個可以換成 users.find，會比較好讀一點

要用 Promise 改寫也行，那個確實在這個課不會提到，那可以自己想想看怎麼改，也可以從 XHR 改成用 fetch

**Reply by a1740942002** [reply]
Hi 老師，過了一個月之後我發現已經忘記之前寫的那一坨是在幹嘛了，覺得之前的寫法很難懂也很難修改、維護。

所以這次我用新學到的 Promise 跟 MVC 的概念，對這個專案做了一些改寫。

想請問老師，我這個版本的拆分邏輯相較於上一版，Code 的可讀性是否有提高呢？
( 我自己是覺得用物件導向跟 MVC 我比較好理解，但不知道其它人看起來如何 )

另外也想請問老師，這個版本還有優化的方向嗎？

1. [新版本原始碼](https://github.com/a1740942002/remaster_twitch_app)
2. [新版本展示](https://a1740942002.github.io/remaster_twitch_app/)

希望老師能撥冗查看！感謝

**Reply by aszx87410** [reply]
我倒是覺得上一個寫法滿 ok 的，小的專案用小的寫法看 code 起來比較容易，我只要看一個檔案就可以知道整個程式在幹嘛，分開之後我可能要看三四個檔案才能知道每一個細節

新版有個地方有寫錯：`window.onload = getStreams();`，應該是 `window.onload = getStreams`

至於程式碼怎麼拆分跟怎麼寫，我是建議等之後學到 react/vue/angular 再來想，因為從那幾個的寫法中有很多值得借鏡的地方

**Reply by a1740942002** [reply]
感謝老師回覆！

關於那個 `window.onload` 我也覺得很奇怪，我原本的理解也是寫 `getStreams`，但是會報錯；寫 `getStreams()` 卻不會，之後再來找時間研究看看。

另外其實這次改寫成 MVC，是因為最近剛學完 Laravel，覺得 PHP 的物件導向相較於 JavaScript 較為嚴謹跟有結構，所以就來改寫試試看了，算是借鑒了 Laravel 這個後端框架的寫法。

因為我當初是先學 [BE201] 課程中的 Express，也是 MVC 的架構，結果發現寫一寫就不知道現在這個變數是什麼了，每個都長得很像，直到我跑去學了 Laravel，現在回頭再來看 Express 才發現比較好理解了。

之前在 [FE301] 中的前幾堂課有理解到 React 的精髓 ( state => UI )，也有問過老師 Vue 是如何實作的，但目前還沒抓到精髓，只知道 Vue 2核心好像是用到 `Object.definProperty` 來實現雙向綁定，所以在這個專案也有用 `Object.definProperty` 來實作無限滾軸的機制。

這兩天應該補完 Webpack 的部分就會全心鑽研框架的使用了，其實我一開始很早就碰框架，只是學一學都不知道為什麼可以這樣做，所以這陣子就回過頭來重新把基礎打完，想說之後重回框架應該會事半功倍。

最後想請問老師，請問 Design Pattern 大概是什麼程度才需要學呢？有稍微看了一些教學但真的無法理解那是用來幹嘛的。

**Reply by aszx87410** [reply]
那應該就是回來把基礎打好之後，去複習一下物件導向，然後再去寫框架，再去碰 design pattern  
理由一樣，框架裡面可以看到比較多設計模式的應用
